### PR TITLE
Migrate from slacker to slack-sdk

### LINKED ISSUE #209: Any plan to change to rtm.connect() ?
I used this bot for a while, its an awesome project, however not usable with Slack Enterprise as it requires rtm.connect()

### LINKED ISSUE #210: response or reply with block message?
I discovered from slack api that can be using rich messaging called "block".
https://api.slack.com/messaging/composing/layouts

I tried to send using `send_webapi()`, but it fails.
Is there a way to send it?

```
msg={
  "channel": "C1H9RESGL",
  "blocks": [
    {
      "type": "section",
      "text": {
        "type": "mrkdwn",
        "text": "Danny Torrence left the following review for your property:"
      }
    },
    {
      "type": "section",
      "block_id": "section567",
      "text": {
        "type": "mrkdwn",
        "text": "<https://google.com|Overlook Hotel> \n :star: \n Doors had too many axe holes, guest in room 237 was far too rowdy, whole place felt stuck in the 1920s."
      },
      "accessory": {
        "type": "image",
        "image_url": "https://is5-ssl.mzstatic.com/image/thumb/Purple3/v4/d3/72/5c/d3725c8f-c642-5d69-1904-aa36e4297885/source/256x256bb.jpg",
        "alt_text": "Haunted hotel image"
      }
    },
    {
      "type": "section",
      "block_id": "section789",
      "fields": [
        {
          "type": "mrkdwn",
          "text": "*Average Rating*\n1.0"
        }
      ]
    }
  ]
}

message.send_webapi('',json.dumps(msg)) # fails with slacker.Error: no_text
message.send_webapi('test',json.dumps(msg)) # I only get 'test' text in reply.

```

### LINKED ISSUE #227: rtm.start response is changing - appears to be a breaking one
According to a [new entry from slack](https://api.slack.com/changelog/2021-10-rtm-start-to-stop) it appears that on September 20th, 2022 the `rtm.start()` response will be truncated and become identical to the `rtm.connect()` response [see docs](https://api.slack.com/methods/rtm.connect#responses)

This response will be missing the `users`, `channels`, `groups`, and `ims` keys that this library uses [on initialization of a new client](https://github.com/scrapinghub/slackbot/blob/2ffb3605e12bb62a8f085de22e827ea2c58e2d9f/slackbot/slackclient.py#L60-L67) 

I believe these changes will break this client on that date. Is my analysis correct and if so, are there any plans to update this library to survive these changes?

### PR DESCRIPTION
**Pull request by jacobtruman**
A newly created legacy bot token fails to authenticate using the older auth method used by the `slacker` module. The `rtm.start` method has also been deprecated, so migrated to `rtm.connect`. The `slack-sdk` module also provides blocks support.

Resolves #227.
Resolves #210.
Resolves #209.

### CODE REVIEW

**Comment by amuraru** [comment]
+1

File: requirements.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 requests>=2.4.0
 websocket-client>=0.22.0,<=0.44.0
-slacker>=0.9.50
+slack-sdk>=3.13.0
```
Comment (amuraru) [inline requirements.txt:3]: update https://github.com/scrapinghub/slackbot/blob/develop/setup.py#L9 as well

Comment (jacobtruman) [reply requirements.txt:3]: good call - this has now been updated

File: slackbot/slackclient.py [hunk 2]
```diff
@@ -61,10 +61,11 @@ def parse_slack_login_data(self, login_data):
...
         self.username = self.login_data['self']['name']
-        self.parse_user_data(login_data['users'])
-        self.parse_channel_data(login_data['channels'])
-        self.parse_channel_data(login_data['groups'])
-        self.parse_channel_data(login_data['ims'])
+        self.parse_user_data(self.webapi.users_list()['members'])
+        self.parse_channel_data(self.webapi.conversations_list(
+            exclude_archived=True,
+            types="public_channel,private_channel"
+        )['channels'])
```
Comment (amuraru) [inline slackbot/slackclient.py]:
```suggestion
        for page in self.webapi.users_list(limit=200):
            self.parse_user_data(page['members'])
        for page in self.webapi.conversations_list(exclude_archived=True,
                                                   types="public_channel,private_channel",
                                                   limit=200):
            self.parse_channel_data(page['channels'])
```

to iterate through all pages

**Comment by amuraru** [comment]
@jacobtruman this is promising. I tested it with my own bot and encountered an issue though:

the bot enters an infinite loop due to rtm websocket disconnect:

```
2021-12-23 19:41:11,009 - slackbot.slackclient - WARNING - reconnected to slack rtm websocket
2021-12-23 19:41:16,020 - slackbot.slackclient - WARNING - lost websocket connection, try to reconnect now
```

Should we look into using the RTMClient instead?
https://slack.dev/python-slack-sdk/real_time_messaging.html#real-time-messaging-rtm

File: slackbot/slackclient.py (hunk 2 above)
Comment (jacobtruman) [reply slackbot/slackclient.py:68]: @amuraru you are correct - we should implement pagination, though I believe we need to use the `cursor` parameter, passing the `next_cursor` value obtained from the `response_metadata` - I will work on this

**Comment by jacobtruman** [comment]
> @jacobtruman this is promising. I tested it with my own bot and encountered an issue though:
> 
> the bot enters an infinite loop due to rtm websocket disconnect:
> 
> ```
> 2021-12-23 19:41:11,009 - slackbot.slackclient - WARNING - reconnected to slack rtm websocket
> 2021-12-23 19:41:16,020 - slackbot.slackclient - WARNING - lost websocket connection, try to reconnect now
> ```
> 
> Should we look into using the RTMClient instead? https://slack.dev/python-slack-sdk/real_time_messaging.html#real-time-messaging-rtm

@amuraru good catch - is this something easily duplicated, so that I can properly test it?

File: slackbot/slackclient.py (hunk 2 above)
Comment (jacobtruman) [reply slackbot/slackclient.py]: @amuraru I implemented pagination, but slightly differently, to account for rate limit errors

File: slackbot/slackclient.py [hunk 3]
```diff
@@ -35,16 +37,15 @@ def __init__(self, token, timeout=None, bot_icon=None, bot_emoji=None, connect=T
         self.rtm_start_args = rtm_start_args
 
         if timeout is None:
-            self.webapi = slacker.Slacker(self.token)
+            self.webapi = slack_sdk.WebClient(self.token)
         else:
-            self.webapi = slacker.Slacker(self.token, timeout=timeout)
+            self.webapi = slack_sdk.WebClient(self.token, timeout=timeout)
```
Comment (amuraru) [inline slackbot/slackclient.py:42]:
```suggestion
            self.webapi = slack_sdk.WebClient(self.token, timeout=timeout)
        # This handler does retries when HTTP status 429 is returned
        from slack_sdk.http_retry.builtin_handlers import RateLimitErrorRetryHandler
        rate_limit_handler = RateLimitErrorRetryHandler(max_retry_count=100)
        # Enable rate limited error retries as well
        self.webapi.retry_handlers.append(rate_limit_handler)
```

I'd recommend we register the builtin `RateLimitErrorRetryHandler` here

Comment (jacobtruman) [reply slackbot/slackclient.py:42]: done

**Comment by jacobtruman** [comment]
> @jacobtruman this is promising. I tested it with my own bot and encountered an issue though:
> 
> the bot enters an infinite loop due to rtm websocket disconnect:
> 
> ```
> 2021-12-23 19:41:11,009 - slackbot.slackclient - WARNING - reconnected to slack rtm websocket
> 2021-12-23 19:41:16,020 - slackbot.slackclient - WARNING - lost websocket connection, try to reconnect now
> ```
> 
> Should we look into using the RTMClient instead? https://slack.dev/python-slack-sdk/real_time_messaging.html#real-time-messaging-rtm

@amuraru I was able to duplicate this issue and added a ping to the parse user and channel data methods to prevent websocket disconnect. I am looking into implementing using RTMClient, but it looks like it might be a pretty hefty rewrite - I still think it is a good idea and am still experimenting with it.

**Comment by jacobtruman** [comment]
closing in favor of #229 under feature branch

**Pull request closed without merging**